lib/tbot/workloadidentity/workloadattest/systemd_linux.go

```diff
@@ -0,0 +1,83 @@
+//go:build linux
+
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package workloadattest
+
+import (
+	"context"
+	"log/slog"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-systemd/v22/dbus"
+	"github.com/gravitational/trace"
+
+	workloadidentityv1pb "github.com/gravitational/teleport/api/gen/proto/go/teleport/workloadidentity/v1"
+)
+
+// SystemdAttestor attests the identity of a Systemd service.
+type SystemdAttestor struct {
+	cfg    SystemdAttestorConfig
+	log    *slog.Logger
+	dialer func(context.Context) (dbusConn, error)
+}
+
+// NewSystemdAttestor creates a SystemdAttestor with the given configuration.
+func NewSystemdAttestor(cfg SystemdAttestorConfig, log *slog.Logger) *SystemdAttestor {
+	return &SystemdAttestor{
+		dialer: func(ctx context.Context) (dbusConn, error) {
+			return dbus.NewWithContext(ctx)
+		},
+	}
+}
+
+// Attest the identity of the given systemd workload.
+func (a *SystemdAttestor) Attest(ctx context.Context, pid int) (*workloadidentityv1pb.WorkloadAttrsSystemd, error) {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	conn, err := a.dialer(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err, "creating dbus connection")
+	}
+	defer conn.Close()
+
+	unit, err := conn.GetUnitNameByPID(ctx, uint32(pid))
+	if err != nil {
+		return nil, trace.Wrap(err, "getting unit name")
+	}
```

Comment:
Should we handle different error types?
Eg, lacking permissions vs PID does not exist

Reply:
Maybe, yeah! I'm just not totally sure of the most useful way to handle them.

For example, we could check for the dbus [`BUS_ERROR_NO_UNIT_FOR_PID`](https://github.com/systemd/systemd/blob/457f4cfbbae4b82a291b2994e0afb631cabd913f/src/libsystemd/sd-bus/bus-common-errors.h#L8) error code - but I think we'd still just want to return the error to the caller (as is the convention for any failing attestor). We could change the error message, but the [existing error message](https://github.com/systemd/systemd/blob/4d44670dc615e254703cff6a199df19e7f741b98/src/core/dbus-manager.c#L562) is already quite clear:

> 2025/03/06 19:27:32 PID 444 does not belong to any loaded unit.